File: 8.回溯/8.组合总和II/40.组合总和-ii.cpp
/*
 * @lc app=leetcode.cn id=40 lang=cpp
 *
 * [40] 组合总和 II
 */
#include "tools.h"
// @lc code=start
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& c, int t) {
        ranges::sort(c);
        vector<vector<int>> res;
        vector<int> pack;
        vector<bool> vis(c.size(), 0);
        dfs(0, t, res, pack, 0, c, vis);
        return res;
    }
    void dfs(int sums, int t, vector<vector<int>>& res, vector<int>& pack, int sta, vector<int>& c, vector<bool>& vis)
    {
        if (sums == t)
        {
            res.emplace_back(pack);
            return;
        }
        for (int i = sta; i < c.size(); i ++ )
        {
            if (i > sta && c[i] == c[i - 1] && vis[i] == 0) continue;
            sums += c[i];
            vis[i] = 1;
            if (sums > t)
            {
                sums -= c[i];
                break;
            }
            pack.push_back(c[i]);
            dfs(sums, t, res, pack, i + 1, c, vis);
            pack.pop_back();
            vis[i] = 0;
            sums -= c[i];
        }
    }
};
// @lc code=end

File: 8.回溯/8.组合总和II/40.组合总和-ii.py
#
# @lc app=leetcode.cn id=40 lang=python3
#
# [40] 组合总和 II
#
from mytools import *
# @lc code=start
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        vis = [0] *  len(candidates)
        def dfs(sta, sums, res: List, pack: List):
            if sums == target:
                if pack not in res:
                    res.append(pack.copy())
                return
            for i in range(sta, len(candidates)):
                if i > sta and candidates[i - 1] == candidates[i] and vis[i - 1] == 0: continue
                sums += candidates[i]
                if sums > target:
                    sums -= candidates[i]
                    break
                pack.append(candidates[i])
                vis[i] = 1
                dfs(i + 1, sums, res, pack)
                sums -= candidates[i]
                vis[i] = 0
                pack.pop()
        res = []
        candidates.sort()
        dfs(0, 0, res, [])
        return res
# @lc code=end

